hw/chisel/src/test/scala/snax/csr_manager/TestUtils.scala

```diff
@@ -0,0 +1,64 @@
+package snax.csr_manager
+
+import chisel3._
+import org.scalatest.flatspec.AnyFlatSpec
+import chiseltest._
+import scala.math.BigInt
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.Tag
+import scala.util.control.Breaks._
+
+trait HasCsrManagerTestUtils {
+
+  // write csr helper function
+  def write_csr[T <: CsrManager](dut: T, addr: Int, data: Int) = {
+
+    // give the data and address to the right ports
+    dut.io.csr_config_in.req.bits.write.poke(1.B)
+    dut.io.csr_config_in.req.bits.data.poke(data.U)
+    dut.io.csr_config_in.req.bits.addr.poke(addr.U)
+    dut.io.csr_config_in.req.valid.poke(1.B)
+
+    // wait for grant
+    while (dut.io.csr_config_in.req.ready.peekBoolean() == false) {
+      dut.clock.step(1)
+    }
+
+    dut.clock.step(1)
```

Comment:
I think this does double clock when waiting for ready? Like:

```scala

if (dut.io.csr_config_in.req.ready.peekBoolean() == false) {
    while (dut.io.csr_config_in.req.ready.peekBoolean() == false) {
          dut.clock.step(1)
    }
} else {
    dut.clock.step(1)
}

```
The point is just to avoid double clocking haha!

And just a minor nitpick but would it be better to test for both valid and ready? I know valid was set prior (I mean duh hahaha) but like just to make sure when someone reads the code they understand you meant to check the valid-ready response.

Reply:
No, the timing is correct. Because the while loop will exit on the immediate cycle when ready is high. Therefore it does not do double clock before doing the `valid.poke`.